code/game/objects/structures/barrels/cask_rack.dm

```diff
@@ -0,0 +1,146 @@
+
+/obj/structure/cask_rack/proc/adjust_barrel_offsets(atom/movable/barrel, barrel_position)
+	barrel.reset_offsets(anim_time = 0)
+	barrel.vis_flags |= (VIS_INHERIT_LAYER | VIS_INHERIT_PLANE)
+
+/obj/structure/cask_rack/Entered(atom/movable/AM, atom/old_loc)
+	. = ..()
+	if(istype(AM) && !QDELETED(AM) && is_type_in_list(AM, get_stackable_barrel_types()))
+		vis_contents |= AM
+		recalculate_barrel_values()
+
+/obj/structure/cask_rack/Exited(atom/movable/AM, atom/new_loc)
+	. = ..()
+	if(istype(AM) && is_type_in_list(AM, get_stackable_barrel_types()))
+		vis_contents -= AM
+		AM.vis_flags = initial(AM.vis_flags)
+		AM.reset_offsets(anim_time = 0)
+		recalculate_barrel_values()
+
+/obj/structure/cask_rack/proc/recalculate_barrel_values()
+	if(length(contents))
+		density     =  TRUE
+		anchored    =  TRUE
+		obj_flags  &= ~OBJ_FLAG_ANCHORABLE
+		atom_flags |=  ATOM_FLAG_CLIMBABLE
+	else
+		density     =  FALSE
+		obj_flags  |=  OBJ_FLAG_ANCHORABLE
+		atom_flags &= ~ATOM_FLAG_CLIMBABLE
+	update_icon()
+
+/obj/structure/cask_rack/proc/try_unstack_barrel(atom/movable/barrel, turf/target, mob/user)
+	if(!loc)
+		return FALSE
+	if(!barrel)
+		if(!length(contents))
+			to_chat(user, SPAN_WARNING("\The [src] has nothing stacked on it."))
+			return FALSE
+		barrel = contents[length(contents)]
+	if(!istype(barrel) || !barrel.simulated)
+		return FALSE
+	if(target && (!isturf(target) || !loc.Adjacent(target))) // TODO: Enter() or CanPass() checks instead of relying on step_towards() below.
+		to_chat(user, SPAN_NOTICE("You cannot move \the [barrel] to \the [target]."))
+		return FALSE
+	to_chat(user, SPAN_NOTICE("You move \the [barrel] off \the [src]."))
+	barrel.dropInto(loc)
+	if(target)
+		step_towards(barrel, target)
+	return TRUE
+
+/obj/structure/cask_rack/proc/try_stack_barrel(atom/movable/barrel, mob/user)
+	if(!istype(barrel) || !barrel.simulated || barrel.anchored)
+		return FALSE
+	if(length(contents) >= max_stack)
+		to_chat(user, SPAN_WARNING("\The [src] is already stacked to capacity."))
+		return FALSE
+	var/list/stackable_types = get_stackable_barrel_types()
+	if(!is_type_in_list(barrel, stackable_types))
+		to_chat(user, SPAN_WARNING("\The [src] is cannot hold \the [barrel]."))
+		return FALSE
+	barrel.forceMove(src)
+	to_chat(user, SPAN_NOTICE("You stack \the [barrel] onto \the [src]."))
+	return TRUE
+
+/obj/structure/cask_rack/proc/get_stackable_barrel_types()
+	var/static/list/_stackable_barrel_types = list(
+		/obj/structure/reagent_dispensers/barrel/cask
+	)
+	return _stackable_barrel_types
+
+// A larger stack, used to arrange up to three casks.
+/obj/structure/cask_rack/large
+	desc      = "A flat rack used to stop casks from rolling around"
+	max_stack = 3
+	w_class   = ITEM_SIZE_LARGE_STRUCTURE
+	icon      = 'icons/obj/structures/barrels/cask_rack_large.dmi'
+
+// We want 'large wooden cask rack' not 'wooden large cask rack'
+/obj/structure/cask_rack/large/update_material_name(override_name)
```

Comment:
kind of feel like we should have some sort of standardized way to do name prefixes at some point...